### ISSUE TITLE
native docker fails with failed to start daemon: Devices cgroup isn't mounted

### ISSUE THREAD

**Issue by farcaller** [nix-community/NixOS-WSL#293]
## Bug description
Docker (native) fails to start with: `failed to start daemon: Devices cgroup isn't mounted`

## To Reproduce
sytemctl start docker

**Expected behavior**
Docker starts

## Logs
```
Sep 21 12:25:50 wsl systemd[1]: Starting Docker Application Container Engine...
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.609487100Z" level=info msg="Starting up"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.610154472Z" level=info msg="libcontainerd: started new containerd process" pid=226085
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.610194431Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.610201965Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.610216123Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.610222596Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50Z" level=warning msg="containerd config version `1` has been deprecated and will be removed in containerd v2.0, please switch to version `2`, see https://github.com/containerd/containerd/blob/main/docs/PLUGINS.md#version-header"
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.626889488Z" level=info msg="starting containerd" revision=v1.7.5 version=v1.7.5
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.637932232Z" level=info msg="loading plugin \"io.containerd.content.v1.content\"..." type=io.containerd.content.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.637993583Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.aufs\"..." type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.638754110Z" level=info msg="skip loading plugin \"io.containerd.snapshotter.v1.aufs\"..." error="aufs is not supported (modprobe aufs failed: exit status 1 \"modprobe: FATAL: Module aufs not found in directory /lib/modules/5.15.123.1-microsoft-standard-WSL2\\n\"): skip plugin" type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.638790722Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.btrfs\"..." type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.638930888Z" level=info msg="skip loading plugin \"io.containerd.snapshotter.v1.btrfs\"..." error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs (ext4) must be a btrfs filesystem to be used with the btrfs snapshotter: skip plugin" type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.638963061Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.devmapper\"..." type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.638976448Z" level=warning msg="failed to load plugin io.containerd.snapshotter.v1.devmapper" error="devmapper not configured"
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.638983361Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.native\"..." type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.638998481Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.overlayfs\"..." type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639053118Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.zfs\"..." type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639189366Z" level=info msg="skip loading plugin \"io.containerd.snapshotter.v1.zfs\"..." error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.zfs must be a zfs filesystem to be used with the zfs snapshotter: skip plugin" type=io.containerd.snapshotter.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639222582Z" level=info msg="loading plugin \"io.containerd.metadata.v1.bolt\"..." type=io.containerd.metadata.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639237270Z" level=warning msg="could not use snapshotter devmapper in metadata plugin" error="devmapper not configured"
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639244394Z" level=info msg="metadata content store policy set" policy=shared
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639334361Z" level=info msg="loading plugin \"io.containerd.differ.v1.walking\"..." type=io.containerd.differ.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639371865Z" level=info msg="loading plugin \"io.containerd.event.v1.exchange\"..." type=io.containerd.event.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639383287Z" level=info msg="loading plugin \"io.containerd.gc.v1.scheduler\"..." type=io.containerd.gc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639400331Z" level=info msg="loading plugin \"io.containerd.service.v1.introspection-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639409870Z" level=info msg="loading plugin \"io.containerd.service.v1.containers-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639418887Z" level=info msg="loading plugin \"io.containerd.service.v1.content-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639427023Z" level=info msg="loading plugin \"io.containerd.service.v1.diff-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639436552Z" level=info msg="loading plugin \"io.containerd.service.v1.images-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639447053Z" level=info msg="loading plugin \"io.containerd.service.v1.leases-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639455740Z" level=info msg="loading plugin \"io.containerd.service.v1.namespaces-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639463485Z" level=info msg="loading plugin \"io.containerd.service.v1.snapshots-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639471331Z" level=info msg="loading plugin \"io.containerd.runtime.v1.linux\"..." type=io.containerd.runtime.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639508934Z" level=info msg="loading plugin \"io.containerd.runtime.v2.task\"..." type=io.containerd.runtime.v2
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639530597Z" level=info msg="loading plugin \"io.containerd.monitor.v1.cgroups\"..." type=io.containerd.monitor.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639778873Z" level=info msg="loading plugin \"io.containerd.service.v1.tasks-service\"..." type=io.containerd.service.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639823210Z" level=info msg="loading plugin \"io.containerd.grpc.v1.introspection\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639836286Z" level=info msg="loading plugin \"io.containerd.internal.v1.restart\"..." type=io.containerd.internal.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639890472Z" level=info msg="loading plugin \"io.containerd.grpc.v1.containers\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639943156Z" level=info msg="loading plugin \"io.containerd.grpc.v1.content\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639954639Z" level=info msg="loading plugin \"io.containerd.grpc.v1.diff\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639963095Z" level=info msg="loading plugin \"io.containerd.grpc.v1.events\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639977944Z" level=info msg="loading plugin \"io.containerd.grpc.v1.healthcheck\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.639995188Z" level=info msg="loading plugin \"io.containerd.grpc.v1.images\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640008785Z" level=info msg="loading plugin \"io.containerd.grpc.v1.leases\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640021510Z" level=info msg="loading plugin \"io.containerd.grpc.v1.namespaces\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640036500Z" level=info msg="loading plugin \"io.containerd.internal.v1.opt\"..." type=io.containerd.internal.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640070005Z" level=info msg="loading plugin \"io.containerd.grpc.v1.snapshots\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640105034Z" level=info msg="loading plugin \"io.containerd.grpc.v1.tasks\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640116286Z" level=info msg="loading plugin \"io.containerd.grpc.v1.version\"..." type=io.containerd.grpc.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640125454Z" level=info msg="loading plugin \"io.containerd.tracing.processor.v1.otlp\"..." type=io.containerd.tracing.processor.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640134462Z" level=info msg="skip loading plugin \"io.containerd.tracing.processor.v1.otlp\"..." error="no OpenTelemetry endpoint: skip plugin" type=io.containerd.tracing.processor.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640141686Z" level=info msg="loading plugin \"io.containerd.internal.v1.tracing\"..." type=io.containerd.internal.v1
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640153540Z" level=error msg="failed to initialize a tracing processor \"otlp\"" error="no OpenTelemetry endpoint: skip plugin"
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640305338Z" level=info msg=serving... address=/var/run/docker/containerd/containerd-debug.sock
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640368653Z" level=info msg=serving... address=/var/run/docker/containerd/containerd.sock.ttrpc
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640413811Z" level=info msg=serving... address=/var/run/docker/containerd/containerd.sock
Sep 21 12:25:50 wsl dockerd[226085]: time="2023-09-21T12:25:50.640424282Z" level=info msg="containerd successfully booted in 0.014193s"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.644414066Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.644447381Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.644462671Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.644469374Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.645329876Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.645356920Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.645367270Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.645372390Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.655163784Z" level=info msg="[graphdriver] using prior storage driver: overlay2"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.662868355Z" level=warning msg="Your kernel does not support cgroup memory limit"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.662900428Z" level=warning msg="Unable to find cpu cgroup in mounts"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.662906670Z" level=warning msg="Unable to find blkio cgroup in mounts"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.662910137Z" level=warning msg="Unable to find cpuset cgroup in mounts"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.662912952Z" level=warning msg="Unable to find pids cgroup in mounts"
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.663289102Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=plugins.moby
Sep 21 12:25:50 wsl dockerd[226064]: time="2023-09-21T12:25:50.663323254Z" level=info msg="stopping healthcheck following graceful shutdown" module=libcontainerd
Sep 21 12:25:51 wsl dockerd[226064]: failed to start daemon: Devices cgroup isn't mounted
Sep 21 12:25:51 wsl systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
```

```
$ mount|grep cgroup
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,size=4096k,nr_inodes=1024,mode=755)
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,name=systemd)
```

nixpkgs: `github:NixOS/nixpkgs/970a59bd19eff3752ce552935687100c46e820a5`
nixos-wsl: `github:nix-community/NixOS-WSL/e7d93d0f478b6fbb47c00d03449dc3d08b90abb7`

**Comment by SuperSandro2000** [comment]
> $ mount|grep cgroup
> tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,size=4096k,nr_inodes=1024,mode=755)
> cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
> cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,name=systemd)

Those should be more.

```
 ▶ mount|grep cgroup
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,size=4096k,nr_inodes=1024,mode=755)
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/cpu type cgroup (rw,nosuid,nodev,noexec,relatime,cpu)
cgroup on /sys/fs/cgroup/cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpuacct)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/net_cls type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
cgroup on /sys/fs/cgroup/net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_prio)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,nosuid,nodev,noexec,relatime,hugetlb)
cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
cgroup on /sys/fs/cgroup/rdma type cgroup (rw,nosuid,nodev,noexec,relatime,rdma)
cgroup on /sys/fs/cgroup/misc type cgroup (rw,nosuid,nodev,noexec,relatime,misc)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,name=systemd)
```


Are you using native systemd? if not please try that

**Comment by farcaller** [comment]
I am using native systemd, yes. This is WSL-2.0 fwiw.

**Comment by K900** [comment]
Should be fixed in WSL 2.0.1, please test.

**Comment by SuperSandro2000** [comment]
I assume this was fixed like written in the last comment
